Use generics for the filter property

nox opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17402)
<!-- Reviewable:end -->

nox committed:
Use generics for the filter property

This introduces an additional shadow type for drop-shadow().

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/gecko_bindings/sugar/ns_css_shadow_item.rs, components/style/values/mod.rs, components/style/properties/longhand/effects.mako.rs, components/style/values/specified/mod.rs, components/style/values/animated/effects.rs and 11 more
 * @emilio: components/layout/display_list_builder.rs, components/layout/fragment.rs, components/style/gecko_bindings/sugar/ns_css_shadow_item.rs, components/style/values/mod.rs, components/style/properties/longhand/effects.mako.rs and 14 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style, layout, and gfx code, but no tests are modified. Please consider adding a test!

nox commented:
r? @emilio 
@bors-servo try
https://treeherder.mozilla.org/#/jobs?repo=try&revision=252af3f864523e297ca71f7bc52b53834e01c983

bors-servo commented:
:hourglass: Trying commit e985b62747f99f7aaa9daf783df674ee0c30f260 with merge 99d5097d23a665d462fce3a25be0f2e89775dcbf...

nox commented:
I failed the serialisation and parsing of `drop-shadow(...)`, new geckotry:

https://treeherder.mozilla.org/#/jobs?repo=try&revision=5a1f8230bb198e2f563c8ab848a5ff0161c0e485

nox commented:
Forgot to reject some negative lengths:

https://treeherder.mozilla.org/#/jobs?repo=try&revision=fbd28f7e40f43899b56c423afaade4adbfdc046c

emilio commented on components/style/properties/helpers/animated_properties.mako.rs:
```diff
@@ -3196,105 +3195,32 @@ impl Animatable for IntermediateShadowList {
-                Ok(IntermediateFilter::${func}(try!(
+            (&Filter::${func}(from_value),
+             &Filter::${func}(to_value)) => {
+                Ok(Filter::${func}(try!(
```
Feel free to untry these while you're around this area of the code.

emilio commented on components/style/values/animated/effects.rs:
```diff
@@ -0,0 +1,202 @@
+/// An animated value for a single `filter`.
+pub type Filter = GenericFilter<
+    Angle,
+    // Should be `NumberOrPercentage`.
```
Perhaps `FIXME`? That way we can grep, I sometimes do that :)

emilio replied:
And, perhaps worth to be explicit and mention `<number-percentage>`, since currently `NumberOrPercentage` isn't exactly that afaict.

emilio commented on components/style/values/animated/effects.rs:
```diff
@@ -0,0 +1,202 @@
+    #[cfg(not(feature = "gecko"))]
+    #[inline]
+    fn add_weighted(&self, _other: &Self, _self_portion: f64, _other_portion: f64) -> Result<Self, ()> {
+        match *self {}
```
This is quite nice actually, didn't expect it to compile without throwing `unreachable` at it.

emilio commented on components/style/values/animated/effects.rs:
```diff
@@ -0,0 +1,202 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Computed types for CSS values related to effects.
```
`s/computed/animated`?

emilio commented on components/style/values/animated/mod.rs:
```diff
@@ -0,0 +1,7 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Animated values.
```
perhaps elaborate a bit on why `computed`/`specified` isn't enough?

emilio commented on components/style/values/generics/effects.rs:
```diff
@@ -0,0 +1,92 @@
+    where
+        W: fmt::Write
+    {
+        if let Some((first, rest)) = self.0.split_first() {
```
I've seen you've used this pattern here and

```rust
if is_empty() {
    return write_str("none");
}

for (ref val, i) in list.enumerate() {
    if i != 0 {
        dest.write_str(" ")?;
    }
    // ...
}
```

In other places.

Any reason for that? Perhaps it's worth to standardize, though I guess this eventually can be derived anyway.

emilio commented on components/style/values/specified/effects.rs:
```diff
@@ -0,0 +1,247 @@
+        context: &ParserContext,
+        input: &mut Parser<'i, 't>
+    ) -> Result<Self, ParseError<'i>> {
+        match true {
```
Why do you need this?

We definitely use:

```rust

#[cfg(feature = "gecko")]
{
    // ...
}
```

In other parts of the code.

nox replied:
This is not stable on 1.15 on statements. Feel free to complain to the Gecko people for us to get a stable bump. :)

emilio replied:
We use it on style, I'm pretty sure, let me find a link...

emilio reviewed (approved):
You rejected too many negative lengths, but with looks good modulo that.

nox commented:
I added a comment for the `animated` module and fixed the failing reftests:

https://treeherder.mozilla.org/#/jobs?repo=try&revision=bb5714442e4efec6fd45493d2354fa8979b1d0cc

nox commented:
These Gecko failures aren't mine.

@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 6f4061d has been approved by `emilio`

<!-- @bors-servo r=emilio 6f4061d4adde05ccf036e44cb9b6bef40f38ccb5 -->

bors-servo commented:
:hourglass: Testing commit 6f4061d4adde05ccf036e44cb9b6bef40f38ccb5 with merge fc2c5b7ef4bd7f36b6aa4e0540adc5b8ef3882c0...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7589), [arm32](http://build.servo.org/builders/arm32/builds/7807), [arm64](http://build.servo.org/builders/arm64/builds/7844), [linux-dev](http://build.servo.org/builders/linux-dev/builds/7932), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4689), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4717), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7917), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/400), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/398), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4622), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4684), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/397), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/400), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3165)
Approved by: emilio
Pushing fc2c5b7ef4bd7f36b6aa4e0540adc5b8ef3882c0 to master...
